zipp/__init__.py

```diff
@@ -313,6 +322,47 @@ def iterdir(self):
         subs = map(self._next, self.root.namelist())
         return filter(self._is_child, subs)
 
+    def match(self, path_pattern):
+        return pathlib.Path(self.at).match(path_pattern)
+
+    def is_symlink(self):
+        return False  # See #82102
+
+    @contextlib.contextmanager
+    def _scandir(self):  # Needed for glob
+        yield self.iterdir()
+
+    def _make_child_relpath(self, relpath):  # Needed for glob
+        return self / relpath
+
+    def glob(self, pattern):
+        if not pattern:
```

Comment:
These lines are copied with only trivial modification from the `pathlib` implementation. The trivial modifications are two:

* changing places that do `self._flavour` to instead do `Path(self.at)._flavour`
* at the very end, passing the `zipfile.Path` itself  to `selector.select_from` (so it can do correct IO) rather than passing the temporary `Path(self.at)`

It seems like this could be refactored somehow to share its implementation but that would need to happen on the CPython side with changes to the `pathlib` module?

Reply:
> It seems like this could be refactored somehow to share its implementation but that would need to happen on the CPython side with changes to the `pathlib` module?

Yes, perhaps. Here's how I'd propose to accomplish that:

- Create a new module that contains this common behavior as an abstract base class.
- The ABC would have not only the abstract methods, but also concrete shared behavior.
- Demonstrate the functionality/behavior in zipp (or other third-party library).
- Adopt the ABC in CPython 3.12+.

There's already something like this in [importlib resources Traversable](https://github.com/python/importlib_resources/blob/33719f1864a6bba2b9a26e6ea24e98d84a10721c/importlib_resources/abc.py#L60), which was originally invented to capture the shared functionality of zipfile.Path and pathlib.Path objects on which a user could rely for resources. Importantly, however, it also now serves as the minimum interface that a resource provider needs to supply. In an ideal world, these implementations could converge too.

Reply:
After more consideration, I'm skeptical that the pathlib implementation adds value here. It certainly adds a lot of complication, including factors irrelevant to zip files (`ntpath.sep`), so I'm inclined not to depend on any pathlib functionality until that functionality can be shaped in such a way that it applies cleanly to the concerns here.